### PR TITLE
Fix pg_wire->execute message outbound handling, commandComplete must not flush

### LINKED ISSUE #11357: Observed flakyness when using the "fast_executemany" option with the "DataDirect PostgreSQL ODBC Driver"
Hi there,

we ran into a specific issue when using CrateDB with ODBC, might have identified a flaw and wanted to share the outcome of our investigations with you.

### Problem report
When connecting to CrateDB's PostgreSQL interface using [unixODBC](http://www.unixodbc.org) through its [pyodbc](https://pypi.org/project/pyodbc/) binding and enabling the `fast_executemany` option, the [Progress DataDirect PostgreSQL ODBC Driver](https://www.progress.com/odbc/postgresql) shows flaky communication and synchronization behavior.

Everything works well when using PostgreSQL. Also, when using the vanilla [psqlODBC - PostgreSQL ODBC driver](https://odbc.postgresql.org/), no flaws happened.

### Details
The software versions we are using are:
- CrateDB 4.5.1
- PostgreSQL 13.2
- Docker 20.10.5
- Python 3.9.5
- pyodbc 4.0.30
- Progress DataDirect Connect ODBC PostgreSQL Wire Protocol Driver 7.1.6
- psqlODBC - ODBC driver for PostgreSQL 11.00.0000

### Reproduction
The whole setup for investigating this issue is wrapped into a repository [1] in order to make reproduction effortless. Its README document also outlines the observations in more detail. The setup must be run on Linux or a respective emulated environment because it only includes ODBC drivers for Linux.

A visual representation of the flaky behavior in the specific scenario is attached in form of a screenshot capturing the outcome of the test suite.

### Thoughts
Maybe this is related to what we investigated on behalf of https://github.com/brianc/node-postgres/issues/2454 and https://github.com/brianc/node-postgres/issues/2455 and mitigated with https://github.com/crate/crate/pull/10979. However, that is really just a wild guess.

With kind regards,
Andreas.

[1] https://github.com/amotl/cratedb-datadirect-odbc

/cc @hammerhead, @proddata, @jayeff 

---

image

### PR DESCRIPTION
**Pull request by seut**
A `CommandComplete` outbound message should not flush, a `ReadyForQuery` is
expected at the end of the message flow which will flush.
Flushing in-between can cause problems with some clients as they will treat
a flush as the end of the message flow response and thus will treat such a
flow as invalid.
The proprietary Progress PgSQL ODBC DataDirect driver for example won't work
otherwise.

Fixes #11357.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: docs/appendices/release-notes/unreleased.rst [hunk 1]
```diff
@@ -97,6 +97,9 @@ Changes
 Fixes
 =====
 
+- Fixed an issue at the PostgreSQL protocol message handling causing some
+  clients/driver to fail on inserts using prepared statements.
```
Comment (amotl) [inline docs/appendices/release-notes/unreleased.rst:101]:
```suggestion
  clients/drivers to fail on inserts using prepared statements.
```

**Review by amotl — APPROVED** [review]
Thank you very much for tracking this down and coming up with a fix!  💯

**Comment by seut** [comment]
Btw. I've run most of the crate/crate-qa client tests against this branch to validate that this change won't break anything.

**Comment by seut** [comment]
retest this please

looks like unrelated? failure:
```
REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=E93656FD24942B59 -Dtests.class=io.crate.integrationtests.StaticInformationSchemaQueryTest -Dtests.method="testGtNumber" -Dtests.locale=lg-UG -Dtests.timezone=Asia/Kamchatka

java.lang.AssertionError: expected:<1> but was:<0>
at __randomizedtesting.SeedInfo.seed([E93656FD24942B59:2CC3DEED136E0E3C]:0)
at org.junit.Assert.fail(Assert.java:88)
at org.junit.Assert.failNotEquals(Assert.java:834)
at org.junit.Assert.assertEquals(Assert.java:645)
at org.junit.Assert.assertEquals(Assert.java:631)
at io.crate.integrationtests.StaticInformationSchemaQueryTest.testGtNumber(StaticInformationSchemaQueryTest.java:178)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:567)
at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.lang.Thread.run(Thread.java:831)
```

**Review by mfussenegger — APPROVED** [review]
Nice :+1: 

Do you think this change might make it possible to remove the `DelayableWriteChannel` altogether? Maybe worth following up on and try it out.

**Comment by seut** [comment]
> Do you think this change might make it possible to remove the DelayableWriteChannel altogether? Maybe worth following up on and try it out.

I'll try this out.


On the other hand it seems my patch breaks the npsql driver (didn't catch this before due to my fault), so this needs some more iteration anyway unfortunately :(

**Comment by mfussenegger** [comment]
Closing this for now - see https://github.com/crate/crate/issues/11357#issuecomment-934340829

**Pull request closed without merging**